### PR TITLE
codegen graphql schema

### PR DESCRIPTION
**Pull request by Tobbe**
Use `codegen` instead of programmatically calling the CLI.

This sets us up to better handle #4762
If/when #4849 goes in we could add a link to that when printing the error message

Easiest way to test this PR is to use [the Gitpod link](https://gitpod.io/#https://github.com/redwoodjs/redwood/pull/5213) and then paste the Book and Shelf models below into schema.prisma
```
model Book {
  id      Int    @id @default(autoincrement())
  title   String @unique
  Shelf   Shelf? @relation(fields: [shelfId], references: [id])
  shelfId Int?
}

model Shelf {
  id    Int    @id @default(autoincrement())
  name  String @unique
  books Book[]
}
```

After that just run `yarn rw g sdl Book` and you'll see an error message printed

### CODE REVIEW

**Comment by orta** [comment]
Note that this conflicts pretty heavily with #5216 which right now needs to evaluate code before running

**Comment by Tobbe** [comment]
@orta This PR mainly touches graphqlSchema.ts and yours is mainly in graphqlCodeGen.ts. Can you please help me understand how they conflict? Thanks!

**Comment by orta** [comment]
oh good point - it doesn't!

File: packages/internal/src/generate/graphqlSchema.ts [hunk 1]
```diff
@@ -1,59 +1,102 @@
...
+      const schemaPrisma = fs.readFileSync(getPaths().api.dbSchema)
+
+      console.error('Schema loading failed.', e.message)
+
+      if (name && schemaPrisma.includes(`model ${name}`)) {
+        // Not all SDLs need to be backed by a DB model, but if they are we can
+        // provide a more helpful error message
+
+        console.error('')
+        console.error(
```
Comment (Tobbe) [inline packages/internal/src/generate/graphqlSchema.ts:68]: @dthyresson Can you please help me out with what kind of messaging we want here?

Comment (dthyresson) [reply packages/internal/src/generate/graphqlSchema.ts:68]: Hi @Tobbe ideally we'd share some language and link to this doc: https://github.com/redwoodjs/redwood/pull/4849/files

But waiting on @jtoar or others to review and publish.

> If you see `Error: Unknown type: ...`, don't panic!
> It's a known issue with GraphQL type generation.
> It happens when you generate the SDL of a Prisma model that has relations **before the SDL for the related model** exists.
> Please see [How to Fix Type Generation Error with Relations​](./schema-relations#errors-when-generating-sdl-or-scaffolds-for-relations) for help.

**Review by dthyresson — CHANGES_REQUESTED** [review]
@Tobbe I added some language from the docs to help resolve the SDL generator type issue.

I think similar language there and a link to the docs when published can be helpful.

File: packages/internal/src/generate/graphqlSchema.ts (hunk 1 above)
Comment (jtoar) [reply packages/internal/src/generate/graphqlSchema.ts:68]: @Tobbe I just merged https://github.com/redwoodjs/redwood/pull/4849

**Comment by Tobbe** [comment]
This is what it looks like now. I think I'll add a couple of newlines 
image

File: packages/internal/src/generate/graphqlSchema.ts (hunk 1 above)
Comment (Tobbe) [reply packages/internal/src/generate/graphqlSchema.ts:70]: @jtoar Thanks. I added a link to the terminal output

**Comment by Tobbe** [comment]
This is what it looks like now, and I'm happy with that.

image

**Comment by dthyresson** [comment]
> This is what it looks like now, and I'm happy with that.
> 
> image
> 
> In my own terminal the link looks better
> 
> image

This is great @Tobbe . Leaving to @jtoar to finalize language.

**Comment by jtoar** [comment]
@Tobbe Awesome! This is gonna save us a ton of Discord threads. 😄

I have two suggestions:

First, I'm worried the `Maybe you need to generate SDL or scaffolding for Book first` part may make users think that they need to start over (by `git clean`ing, etc), generate the other model (here, `Book`), only to find another error message saying `Maybe you need to generate SDL or scaffolding for Shelf first`.

So I propose we change it to:

```
It looks like you have a Shelf model in your Prisma schema.
If it's part of a relation, you may have to generate SDL or scaffolding for Shelf too.
If you haven't done so yet, ignore the error message here and do so now.
```

My reasoning is that you have to generate one before the other, so don't tell them to start over, just tell them to generate the other one. You could comment out the relation fields, generate the SDLs/scaffolds, comment the relations fields back in, and force generate the SDLs/scaffolds, but that's a few too many steps for me when all I need to do is run the other SDL/scaffold generator. But thoughts?

Second, I feel like the message should stand out more. Maybe we could style it a bit to make it more eye catching?

image

This is the code I added (along with a chalk import—it's already a dependency of this package) to get that message; happy to commit it if you agree:

```js
        console.error(
          [
            `  ${chalk.bgYellow(` ${chalk.black.bold('Heads up')} `)}`,
            '',
            chalk.yellow(
              `  It looks like you have a ${name} model in your Prisma schema.`
            ),
            chalk.yellow(
              `  If it's part of a relation, you may have to generate SDL or scaffolding for ${name} too.`
            ),
            chalk.yellow(
              `  If you haven't done so yet, ignore the error message here and do so now.`
            ),
            '',
            chalk.yellow(
              `  See the ${terminalLink(
                'Troubleshooting Generators',
                'https://redwoodjs.com/docs/schema-relations#troubleshooting-generators'
              )} section in our docs for more help.`
            ),
            '',
          ].join('\n')
        )
```

**Comment by Tobbe** [comment]
Yes! I ❤️  this Dom! This is so much better than what I had.

What terminal/shell do you have? Why don't you get the nice looking link I get? (iTerm2 + zsh)

To keep iterating on the text a little bit, what do you think of

```
It looks like you have a Shelf model in your Prisma schema.
If Book and Shelf form a relation, you may have to generate SDL or scaffolding for Shelf too.
So, if you haven't done that yet, ignore this error message and run that generator now.
```

**Comment by jtoar** [comment]
> What terminal/shell do you have? Why don't you get the nice looking link I get? (iTerm2 + zsh)

I use the default mac terminal and fish shell. I think the mac terminal is unsupported. It's ok!

I like the copy! If I could suggest one more thing, I'd change `If Book and Shelf form a relation` to `If Book and Shelf have a relation`.

I'll add the commit now. There's one thing I don't know how to do yet: how do you get "Book"?

File: packages/internal/src/generate/graphqlSchema.ts [hunk 2]
```diff
@@ -68,21 +69,30 @@ export const generateGraphQLSchema = async () => {
...
-          `It looks like you have a ${name} model in your database schema.`
+          [
+            `  ${chalk.bgYellow(` ${chalk.black.bold('Heads up')} `)}`,
+            '',
+            chalk.yellow(
+              `  It looks like you have a ${name} model in your Prisma schema.`
+            ),
+            chalk.yellow(
+              // how to get "Book"?
+              `  If Book and ${name} have a relation, you may have to generate SDL or scaffolding for ${name} too.`
```
[the remark this answers is not in the document]
Comment (jtoar) [reply packages/internal/src/generate/graphqlSchema.ts:80]: We could do `process.argv[process.argv.length - 1]`, but it feels like that assumes a lot / creates dependencies. Maybe I'm just overthinking it though.

[the remark this answers is not in the document]
Comment (jtoar) [reply packages/internal/src/generate/graphqlSchema.ts:80]: I'll update the test after we get this down)

[the remark this answers is not in the document]
Comment (Tobbe) [reply packages/internal/src/generate/graphqlSchema.ts:80]: I agree. It does feel like it's assuming too much. And we'd want to provide a nice error also when someone just manually runs `yarn rw g types`. I'm sorry, for some reason I just thought we had Book available too, but we don't :(

[the remark this answers is not in the document]
Comment (jtoar) [reply packages/internal/src/generate/graphqlSchema.ts:80]: No worries! Should we just rewrite the message so it doesn't need the value of whatever "Book" would be?

**Comment by jnhooper** [comment]
Hi! sorry to but into this conversation, this all looks great and i think it'll solve an issue i was butting my head against. I was trying to look through this pr to see what your solution looked like but i couldn't find the successful test fixture, so i was hoping someone here could take a quick peek/ this could be illuminating for anyone else coming across this pr like me before this gets released.
```
model Item {
  id          Int       @id @default(autoincrement())
  name        String
  quantity    Int
    
  createdBy   User      @relation("createdItems", fields: [createdByUserId], references: [id])
  pickedUpBy  User?     @relation("pickedUpItems", fields: [pickedUpById], references: [id])
  

  createdByUserId Int
  pickedUpById    Int

}

model User {
  id                  Int       @id @default(autoincrement())
  email               String    @unique
  name                String
  itemsCreated      Item[]      @relation("createdItems")
  itemsPickedUp     Item[]      @relation("pickedUpItems")
}```
then when scaffolding these i run into the problem `Unknown type User` and `Unknown type Item` in the respective sdl files. I solved this by just manually copying the `types` into each other's sdl files 
```
//item.sdl
export const schema = gql`
# had to copy the below manually
  type User {
    id: Int!
    email: String!
    name: String!
    itemsCreated: [Item]!
    itemsPickedUp: [Item]!
  }

# this was auto generated
  type Item {
    id: Int!
    name: String!
    quantity: Int!
    createdBy: User!
    pickedUpBy: User
    createdByUserId: Int!
    pickedUpById: Int!
  }
```
this works but seems... not ideal. is it possible to import the types from one file into another?

**Comment by jtoar** [comment]
@jnhooper you shouldn't have to import types into files, just run the sdl or scaffold generator for one model after the other while expecting the first model that to have type errors. Does that make sense? So it'd be:

```
yarn rw g scaffold users
# Will report type errors, but still scaffold ok
yarn rw g scaffold items
# No errors
```

File: packages/internal/src/generate/graphqlSchema.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Tobbe) [reply packages/internal/src/generate/graphqlSchema.ts]: I've done that now :)

**Review by dthyresson — APPROVED** [review]

**Pull request merged**